frappe/public/js/frappe/form/save.js

```diff
@@ -130,12 +130,9 @@ frappe.ui.form.save = function (frm, action, callback, btn) {
 						folded = frm.layout.folded;
 					}
 
-					if (
-						is_docfield_mandatory(doc, df) &&
-						!frappe.model.has_value(doc.doctype, doc.name, df.fieldname)
-					) {
+					if (is_docfield_mandatory(doc, df) && is_null(doc[docfield.fieldname])) {
```

Comment:
This isn't valid fix. 

When the child table is mandatory the `doc.fields` will be truthy but if it's empty then it's not really filled. 

The problem here is incorrect syncing between form builder data and Form.

Reply:
@ankush `cstr` will convert empty array to empty string and becomes `false`.  By the way `doctype.fields` was never empty because `title` field was added already.  Depending on the `localStorage` for validation doesn't seem to be good idea. Doc has already all the required data and any validations should happen on `doc`.  If we have to change `form builder`, then we need to keep track of add/remove field and call `frappe.model.add_child` or `clear_doc` etc which is unnecessary.

Reply:
@ssuda This code is used for every doctype, we can't make a fix here because form builder has a bug. Form builder code should be fixed instead.

Reply:
@ankush this fix is nothing to do with form builder. If doc (any doctype or  `DocType`) has all the data for validation, then referring  to `locals` doesn't seem to be good idea. This may break in future also, if something changes with naming.

Reply:
@ssuda `frm.doc` is not the source of truth. See `form.js`, you'll not find it getting modified anywhere.

Reply:
@ankush `frm.doc` is the one we are sending to server (and It is the final doc).  Here is the code after two lines of `check_mandatory`.

<img width="1256" alt="Screenshot 2023-10-23 at 2 50 01 PM" src="https://github.com/frappe/frappe/assets/928627/f5422978-cf06-4fc0-9869-cfd15c6ba242">

Reply:
I don't want to argue over this. Come on, trust me to do the right thing here :sweat_smile: